runtime/compiler/optimizer/abstractinterpreter/AbsInterpreter.cpp

```diff
@@ -0,0 +1,3072 @@
+      if (blockIt1.currentBlock() != _cfg->getStart()->asBlock() && blockIt1.currentBlock() != _cfg->getEnd()->asBlock()) //skip start and exit block
+         {
+         blockStartIndices.push_back(blockIt1.currentBlock()->getBlockBCIndex());
+         }
+      blockIt1.stepForward();
+      }
+
+   blockStartIndices.push_back(_bci.maxByteCodeIndex());
+   std::sort(blockStartIndices.begin(), blockStartIndices.end());
+
+   // set the bytecode size of each block
+   TR::AllBlockIterator blockIt2(_cfg, _comp);
+   while (blockIt2.currentBlock())
+      {
+      if (blockIt2.currentBlock() != _cfg->getStart()->asBlock() && blockIt2.currentBlock() != _cfg->getEnd()->asBlock()) //skip start and exit block
+         {
+         auto lower = std::lower_bound(blockStartIndices.begin(), blockStartIndices.end(), blockIt2.currentBlock()->getBlockBCIndex()); //binary search for the index
+         auto idx = std::distance(blockStartIndices.begin(), lower);
+         blockIt2.currentBlock()->setBlockSize(blockStartIndices[idx + 1] - blockStartIndices[idx]);
+         }
+      blockIt2.stepForward();
+      }
+   }
+
+TR::ValuePropagation* TR::AbsInterpreter::vp()
+   {
+   if (!_vp)
+      {
+      TR::OptimizationManager* manager = comp()->getOptimizer()->getOptimization(OMR::globalValuePropagation);
+      _vp = (TR::ValuePropagation*) manager->factory()(manager);
+      _vp->initialize();
+      }
+   return _vp;
+   }
+
+void TR::AbsInterpreter::setState(TR::Block* block, TR::AbsStackMachineState* state)
+   {
+   auto iter = _blockStateMap.find(block);
+   if (iter != _blockStateMap.end())
+      _blockStateMap.erase(iter);
+
+   _blockStateMap.insert(std::pair<TR::Block*, TR::AbsStackMachineState*>(block, state));
+   }
+
+TR::AbsStackMachineState* TR::AbsInterpreter::getState(TR::Block* block)
+   {
+   auto iter = _blockStateMap.find(block);
+   if (iter == _blockStateMap.end())
+      return NULL;
+
+   return iter->second;
+   }
+
+bool TR::AbsInterpreter::interpret()
+   {
+   if (comp()->getOption(TR_TraceAbstractInterpretation))
+      traceMsg(comp(), "\nStarting to abstract interpret method %s ...\n", _callerMethod->signature(comp()->trMemory()));
+
+   TR::ReversePostorderSnapshotBlockIterator blockIt(_cfg, comp());
+
+   while (blockIt.currentBlock())
+      {
+      bool success = interpretBlock(blockIt.currentBlock(), false, true);
+      if (!success)
+         return false;
+
+
+      blockIt.stepForward();
+      }
+
+   if (comp()->getOption(TR_TraceAbstractInterpretation))
+      traceMsg(comp(), "\nSuccessfully abstract interpret method %s ...\n", _callerMethod->signature(comp()->trMemory()));
+
+   if (comp()->getOption(TR_TraceBISummary))
+      _inliningMethodSummary->trace(comp());
+
+   return true;
+   }
+
+void TR::AbsInterpreter::transferBlockStatesFromPredeccesors(TR::Block* block, bool insideLoop)
+   {
+   /*** Case 1: No predecessors. Do nothing ***/
+   if (block->getPredecessors().size() == 0)
+      {
+      return;
```

Comment:
How are catch blocks handled?

Reply:
Sorry that I'm a little bit confused about your question. Are you asking about catch blocks because they don't have predecessors?